packages/cli/src/deploy/ensureSystems.ts

```diff
@@ -0,0 +1,128 @@
+import { Client, Transport, Chain, Account, Hex, getAddress } from "viem";
+import { writeContract } from "@latticexyz/common";
+import { System, WorldDeploy, worldAbi } from "./common";
+import { ensureContract } from "./ensureContract";
+import { debug } from "./debug";
+import { resourceLabel } from "./resourceLabel";
+import { getSystems } from "./getSystems";
+import { getResourceAccess } from "./getResourceAccess";
+
+export async function ensureSystems({
+  client,
+  worldDeploy,
+  systems,
+}: {
+  readonly client: Client<Transport, Chain | undefined, Account>;
+  readonly worldDeploy: WorldDeploy;
+  readonly systems: readonly System[];
+}): Promise<readonly Hex[]> {
+  const [worldSystems, worldAccess] = await Promise.all([
+    getSystems({ client, worldDeploy }),
+    getResourceAccess({ client, worldDeploy }),
+  ]);
+  const systemIds = systems.map((system) => system.systemId);
+  const currentAccess = worldAccess.filter(({ resourceId }) => systemIds.includes(resourceId));
+  const desiredAccess = systems.flatMap((system) =>
+    system.allowedAddresses.map((address) => ({ resourceId: system.systemId, address }))
+  );
+
+  const accessToAdd = desiredAccess.filter(
+    (access) =>
+      !currentAccess.some(
+        ({ resourceId, address }) =>
+          resourceId === access.resourceId && getAddress(address) === getAddress(access.address)
+      )
+  );
+
+  const accessToRemove = currentAccess.filter(
+    (access) =>
+      !desiredAccess.some(
+        ({ resourceId, address }) =>
+          resourceId === access.resourceId && getAddress(address) === getAddress(access.address)
+      )
+  );
+
+  // TODO: move each system access+registration to batch call to be atomic
```

Comment:
cc @CrazyNorman

there's a few other TODOs for `batchCall` in here but I think this is one place we can add it that has meaningful impact to the deploy functionality vs. just making it faster

Reply:
> cc @CrazyNorman
> 
> there's a few other TODOs for `batchCall` in here but I think this is one place we can add it that has meaningful impact to the deploy functionality vs. just making it faster

Got you. So If I'm getting you right, we can use `batchCall` to do atomic setup, handling each system as a whole, one by one. This will make the deployment steps more modular.

Reply:
Exactly! If we for some reason fail to register a system but add access control, or vice versa, that feels like a weird state to be in. Using `batchCall` here should help keep the operation atomic so if one step in setting up the system fails, the entire system setup fails.